numbat/src/command.rs

```diff
@@ -46,6 +47,104 @@ impl FromStr for CommandKind {
     }
 }
 
+#[derive(Default)]
+pub enum CommandControlFlow {
+    #[default]
+    Normal,
+    Continue,
+    Return,
+}
+
+pub struct CommandContext<'ctx, 'aux, Editor> {
+    pub ctx: &'ctx mut Context,
+    pub editor: &'aux mut Editor,
+    pub session_history: &'aux SessionHistory,
+    pub interactive: bool,
+}
+
+pub struct CommandRunner<Editor> {
+    help: Option<fn() -> CommandControlFlow>,
+    info: Option<fn(&mut Context, &str) -> CommandControlFlow>,
+    list: Option<fn(&Context, Option<ListItems>) -> CommandControlFlow>,
+    clear: Option<fn(&mut Editor) -> CommandControlFlow>,
+    #[allow(clippy::type_complexity)]
+    save: Option<fn(&Context, &SessionHistory, &str, bool) -> CommandControlFlow>,
+    quit: Option<fn() -> CommandControlFlow>,
+}
+
+impl<Editor> Default for CommandRunner<Editor> {
+    fn default() -> Self {
+        Self {
+            help: None,
+            info: None,
+            list: None,
+            clear: None,
+            save: None,
+            quit: None,
+        }
+    }
+}
+
+impl<Editor> CommandRunner<Editor> {
+    pub fn new_all_disabled() -> Self {
+        Self::default()
+    }
+
+    pub fn enable_help(mut self, action: fn() -> CommandControlFlow) -> Self {
+        self.help = Some(action);
+        self
+    }
+
+    pub fn enable_info(mut self, action: fn(&mut Context, &str) -> CommandControlFlow) -> Self {
+        self.info = Some(action);
+        self
+    }
+
+    pub fn enable_list(
+        mut self,
+        action: fn(&Context, Option<ListItems>) -> CommandControlFlow,
+    ) -> Self {
+        self.list = Some(action);
+        self
+    }
+
+    pub fn enable_clear(mut self, action: fn(&mut Editor) -> CommandControlFlow) -> Self {
+        self.clear = Some(action);
+        self
+    }
+
+    pub fn enable_save(
+        mut self,
+        action: fn(&Context, &SessionHistory, &str, bool) -> CommandControlFlow,
+    ) -> Self {
+        self.save = Some(action);
+        self
+    }
+
+    pub fn enable_quit(mut self, action: fn() -> CommandControlFlow) -> Self {
+        self.quit = Some(action);
+        self
+    }
+
+    pub fn run(&self, cmd: Command, args: CommandContext<Editor>) -> CommandControlFlow {
+        let CommandContext {
+            ctx,
+            editor,
+            session_history,
+            interactive,
+        } = args;
+
+        match cmd {
+            Command::Help => self.help.unwrap()(),
+            Command::Info { item } => self.info.unwrap()(ctx, item),
+            Command::List { items } => self.list.unwrap()(ctx, items),
+            Command::Clear => self.clear.unwrap()(editor),
+            Command::Save { dst } => self.save.unwrap()(ctx, session_history, dst, interactive),
+            Command::Quit => self.quit.unwrap()(),
+        }
```

Comment:
Is it possible to somehow have a tighter coupling between the `CommandRunner` and the `SourcelessCommandParser`? (potentially as a new struct "layered" on top)

It feels like there could be a way where the `.unwrap()`s here and the `let is_supported = …` part below could somehow be tied together.

Sorry for the high-level comment. Let me know if it's unclear.

Reply:
You are correct. Made a bunch of changes, I think in line with what you were thinking.

Basically, instead of producing a `Command`, and then checking whether it's enabled, we only produce a `CommandKind`, check whether that's enabled, and then proceed to parse args. This combines the `is_supported` and `unwrap()`s into one logical component (a single `match`).

Reply:
The changes I made amounted to a pile of spaghetti. I realized that the correct way to implement this:

> It feels like there could be a way where the .unwrap()s here and the let is_supported = … part below could somehow be tied together.

is to stick the functions themselves into the `Command` enum.  Then you can only produce a command if you've already got the corresponding function (i.e. it was `Some(fn(...))` and not `None`).'

So now we have two functions: `get_command` that produces such a command (and is the primary test point), and `try_run_command` that runs the command we just got, which has the necessary functions already stored in whichever variant we're running.